--- a/ui/admin.cpp
+++ b/ui/admin.cpp
@@ -174,7 +174,7 @@
     char* buf = (char*)kiwi_imalloc("console", NBUF + SPACE_FOR_NULL);
     int i, n, err;
 
-    char* args[] = { (char*)"/bin/sh", (char*)"--login", NULL };
+    char* args[] = { (char*)"/bin/bash", (char*)"--login", NULL };
     scall("forkpty", (c->console_child_pid = forkpty(&c->master_pty_fd, NULL, NULL, NULL)));
 
     if (c->console_child_pid == 0) { // child
