diff --git a/_COPYRIGHT b/_COPYRIGHT
--- a/_COPYRIGHT
+++ b/_COPYRIGHT
@@ -15,7 +15,7 @@
         Copyright (c) 2011, Jasper de Waard (jasper@designtown.nl)
 
 Homemade GPS Receiver
-Copyright (C) 2013 Andrew Holme, http://www.holmea.demon.co.uk/GPS/Main.htm
+Copyright (C) 2013 Andrew Holme, http://www.aholme.co.uk/GPS/Main.htm
     Kalman filter, SPP solver
         Copyright (c) 2018-2020 Christoph Mayer, DL1CH
     Galileo, QZSS support
diff --git a/support/coroutines.cpp b/support/coroutines.cpp
--- a/support/coroutines.cpp
+++ b/support/coroutines.cpp
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
-// http://www.holmea.demon.co.uk/GPS/Main.htm
+// http://www.aholme.co.uk/GPS/Main.htm
 //////////////////////////////////////////////////////////////////////////
 
 // Copyright (c) 2014-2017 John Seamons, ZL4VO/KF6VO
diff --git a/support/coroutines.h b/support/coroutines.h
--- a/support/coroutines.h
+++ b/support/coroutines.h
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
-// http://www.holmea.demon.co.uk/GPS/Main.htm
+// http://www.aholme.co.uk/GPS/Main.htm
 //////////////////////////////////////////////////////////////////////////
 
 // Copyright (c) 2014-2016 John Seamons, ZL/KF6VO
diff --git a/support/timing.cpp b/support/timing.cpp
--- a/support/timing.cpp
+++ b/support/timing.cpp
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
-// http://www.holmea.demon.co.uk/GPS/Main.htm
+// http://www.aholme.co.uk/GPS/Main.htm
 //////////////////////////////////////////////////////////////////////////
 
 // Copyright (c) 2014-2017 John Seamons, ZL4VO/KF6VO
diff --git a/support/timing.h b/support/timing.h
--- a/support/timing.h
+++ b/support/timing.h
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
-// http://www.holmea.demon.co.uk/GPS/Main.htm
+// http://www.aholme.co.uk/GPS/Main.htm
 //////////////////////////////////////////////////////////////////////////
 
 #ifndef _TIMING_H_
