diff --git a/extensions/FT8/ft8_lib/decode_ft8.cpp b/extensions/FT8/ft8_lib/decode_ft8.cpp
--- a/extensions/FT8/ft8_lib/decode_ft8.cpp
+++ b/extensions/FT8/ft8_lib/decode_ft8.cpp
@@ -2,6 +2,7 @@
 #include "types.h"
 #include "config.h"
 #include "str.h"
+#include "mem.h"
 #include "rx/CuteSDR/datatypes.h"
 #include "rx/rx.h"
 #include "coroutines.h"
@@ -70,6 +71,7 @@
     int rx_chan;
     bool init;
     bool debug;
+    bool freq_sort;
     int freqHz;
    
     ftx_protocol_t protocol;
@@ -231,6 +233,7 @@
 
 static void decode(int rx_chan, const frame_ft8_t* frame, int freqHz)
 {
+    int i;
     decode_ft8_t *ft8 = &decode_ft8[rx_chan];
     const monitor_t* mon = &frame->mon;
 
@@ -244,7 +247,7 @@
     int num_decoded = 0, num_spots = 0;
 
     // Initialize hash table pointers
-    for (int i = 0; i < kMax_decoded_messages; ++i)
+    for (i = 0; i < kMax_decoded_messages; ++i)
     {
         ft8->decoded_hashtable[i] = NULL;
     }
@@ -252,6 +255,9 @@
     // Go over candidates and attempt to decode messages
     bool need_header = true;
     int limiter = 0;
+    typedef struct { int freq; char *chars; } save_deco_t;
+    save_deco_t save_deco[kMax_candidates];
+    int save_i = 0;
 
     for(int pass = 0; pass < 2; ++pass)
     {
@@ -439,7 +445,8 @@
                 ks = kstr_asprintf(ks, " %d.0 ", i3);
             else
                 ks = kstr_asprintf(ks, " 0.%d*", FTX_MSG_N3(&message));
-                
+
+            char *chars;
             if (pskr_ok) {
                 char *call_to, *call_to_s = NULL, *call_de, *call_de_s, *grid_de, *grid_de_s;
                 if (n == 3)
@@ -478,24 +485,33 @@
 
                 if (n == 3) {
                     // call_to call_de grid4
-                    ext_send_msg_encoded(rx_chan, false, "EXT", "chars",
+                    asprintf(&chars,
                         "%s %s%s %s%s %s" NONL, kstr_sp(ks), ft8->debug? "3> ":"", call_to_s, uploaded? GREEN : "", call_de_s, grid_de_s);
                 } else if (n == 4) {
                     // CQ DX call_de grid4
-                    ext_send_msg_encoded(rx_chan, false, "EXT", "chars",
+                    asprintf(&chars,
                         "%s %s%s %s %s%s %s" NONL, kstr_sp(ks), ft8->debug? "4> ":"", f[0], f[1], uploaded? GREEN : "", call_de_s, grid_de_s);
                 } else {
                     // call_to call_de R grid4
-                    ext_send_msg_encoded(rx_chan, false, "EXT", "chars",
+                    asprintf(&chars,
                         "%s %s%s %s%s %s %s" NONL, kstr_sp(ks), ft8->debug? "9> ":"", call_to_s, uploaded? GREEN : "", call_de_s, f[2], grid_de_s);
                 }
                 
                 free(call_to_s); free(call_de_s); free(grid_de_s);
             } else {
-                ext_send_msg_encoded(rx_chan, false, "EXT", "chars",
+                asprintf(&chars,
                     "%s %s%s\n", kstr_sp(ks), ft8->debug? "0> ":"", text);
             }
 
+            if (ft8->freq_sort) {
+                save_deco[save_i].freq = freq_hz;
+                save_deco[save_i].chars = chars;
+                save_i++;
+            } else {
+                ext_send_msg_encoded(rx_chan, false, "EXT", "chars", chars);
+                kiwi_asfree(chars);
+            }
+
             kstr_free(ks);
             if (need_free) for (n = 0; n < 4; n++) free(f[n]);     // NB: free(NULL) is okay
         }
@@ -503,6 +519,15 @@
     }
     CHECK_PADDING(ft8);
 
+    if (ft8->freq_sort) {
+        qsort(&save_deco, save_i, sizeof(save_deco_t), qsort_intcomp);
+        for (i = 0; i < save_i; i++) {
+            char *chars = save_deco[i].chars;
+            ext_send_msg_encoded(rx_chan, false, "EXT", "chars", chars);
+            kiwi_asfree(chars);
+        }
+    }
+
     LOG(LOG_INFO, "Decoded %d messages, callsign hashtable size %d\n", num_decoded, ft8->callsign_hashtable_size);
     if (num_decoded > 0) {
         char *ks = NULL;
@@ -706,3 +731,9 @@
         "-------------------------------------------------------  new freq %.2f mode %s\n",
         freqHz/1e3, ft8->protocol_s);
 }
+
+void decode_ft8_freq_sort(int rx_chan, int freq_sort)
+{
+    decode_ft8_t *ft8 = &decode_ft8[rx_chan];
+    ft8->freq_sort = freq_sort;
+}
diff --git a/extensions/FT8/FT8.h b/extensions/FT8/FT8.h
--- a/extensions/FT8/FT8.h
+++ b/extensions/FT8/FT8.h
@@ -35,6 +35,7 @@
 C_LINKAGE(void decode_ft8_free(int rx_chan));
 C_LINKAGE(void decode_ft8_setup(int rx_chan, int debug));
 C_LINKAGE(void decode_ft8_protocol(int rx_chan, int freqHz, int proto));
+C_LINKAGE(void decode_ft8_freq_sort(int rx_chan, int freq_sort));
 C_LINKAGE(void decode_ft8_samples(int rx_chan, TYPEMONO16 *samps, int nsamps, int freqHz, u1_t *start_test));
 
 bool ft8_update_vars_from_config(bool called_at_init_or_restart);
diff --git a/extensions/FT8/FT8.cpp b/extensions/FT8/FT8.cpp
--- a/extensions/FT8/FT8.cpp
+++ b/extensions/FT8/FT8.cpp
@@ -30,7 +30,7 @@
 	int rx_chan;
 	int run;
 	int proto;
-	bool debug;
+	bool debug, freq_sort;
 	int last_freq_kHz;
 	
 	bool task_created;
@@ -202,6 +202,13 @@
 		return true;
 	}
 
+    int freq_sort;
+	if (sscanf(msg, "SET ft8_freq_sort=%d", &freq_sort) == 1) {
+	    e->freq_sort = freq_sort;
+        decode_ft8_freq_sort(rx_chan, e->freq_sort);
+		return true;
+    }
+
 	float df;
 	n = sscanf(msg, "SET dialfreq=%f", &df);
 	if (n == 1) {
diff --git a/types.h b/types.h
--- a/types.h
+++ b/types.h
@@ -91,6 +91,8 @@
 #endif
 
 #define NOT_FOUND -1
+#define DEBUG_TRUE    true
+#define DEBUG_FALSE    false
 
 #define ARRAY_EL_LEN(x) ((int)sizeof((x)[0]))
 #define ARRAY_LEN(x)    ((int)(sizeof(x) / ARRAY_EL_LEN(x)))
