diff --git a/rx/rx_cmd.cpp b/rx/rx_cmd.cpp
index 679aed2..d8019ae 100644
--- a/rx/rx_cmd.cpp
+++ b/rx/rx_cmd.cpp
@@ -1266,9 +1266,10 @@ bool rx_common_cmd(int stream_type, conn_t* conn, char* cmd) {
                     idx1 = CLAMP(idx1, 0, cur_len - 1);
 
                     // search starting from given index
+                    dx_print_search(true, "DX_ADM_SEARCH_IDENT idx1=%d\n", idx1);
                     dx_t* dp = &cur_list[idx1];
                     for (i = idx1; dp < &cur_list[cur_len]; i++, dp++) {
-                        dx_print_search(true, "DX_ADM_SEARCH_IDENT LOOP1 <%s> <%s> #%d\n", dp->ident_s, ident, i);
+                        dx_print_search(true, "DX_ADM_SEARCH_IDENT LOOP1 %.2f <%s> <%s> #%d\n", dp->freq, dp->ident_s, ident, i);
                         if (strcasestr(dp->ident_s, ident) != NULL) {
                             pos = i;
                             break;
@@ -1279,7 +1280,7 @@ bool rx_common_cmd(int stream_type, conn_t* conn, char* cmd) {
                     if (pos == -1) {
                         dx_t* dp = &cur_list[0];
                         for (i = 0; dp < &cur_list[cur_len]; i++, dp++) {
-                            dx_print_search(true, "DX_ADM_SEARCH_IDENT LOOP2 <%s> <%s> #%d\n", dp->ident_s, ident, i);
+                            dx_print_search(true, "DX_ADM_SEARCH_IDENT LOOP2 %.2f <%s> <%s> #%d\n", dp->freq, dp->ident_s, ident, i);
                             if (strcasestr(dp->ident_s, ident) != NULL) {
                                 pos = i;
                                 break;
@@ -1303,9 +1304,10 @@ bool rx_common_cmd(int stream_type, conn_t* conn, char* cmd) {
                     idx1 = CLAMP(idx1, 0, cur_len - 1);
 
                     // search starting from given index
+                    dx_print_search(true, "DX_ADM_SEARCH_NOTES idx1=%d\n", idx1);
                     dx_t* dp = &cur_list[idx1];
                     for (i = idx1; dp < &cur_list[cur_len]; i++, dp++) {
-                        dx_print_search(dp->notes, "DX_ADM_SEARCH_NOTES LOOP1 <%s> <%s> #%d\n", dp->notes_s, notes, i);
+                        dx_print_search(dp->notes, "DX_ADM_SEARCH_NOTES LOOP1 %.2f <%s> <%s> #%d\n", dp->freq, dp->notes_s, notes, i);
                         if (dp->notes && strcasestr(dp->notes_s, notes) != NULL) {
                             pos = i;
                             break;
@@ -1316,7 +1318,7 @@ bool rx_common_cmd(int stream_type, conn_t* conn, char* cmd) {
                     if (pos == -1) {
                         dx_t* dp = &cur_list[0];
                         for (i = 0; dp < &cur_list[cur_len]; i++, dp++) {
-                            dx_print_search(dp->notes, "DX_ADM_SEARCH_NOTES LOOP2 <%s> <%s> #%d\n", dp->notes_s, notes, i);
+                            dx_print_search(dp->notes, "DX_ADM_SEARCH_NOTES LOOP2 %.2f <%s> <%s> #%d\n", dp->freq, dp->notes_s, notes, i);
                             if (dp->notes && strcasestr(dp->notes_s, notes) != NULL) {
                                 pos = i;
                                 break;
