diff --git a/extensions/FT8/FT8.cpp b/extensions/FT8/FT8.cpp
index a45eef0..ee9ce62 100644
--- a/extensions/FT8/FT8.cpp
+++ b/extensions/FT8/FT8.cpp
@@ -320,15 +320,21 @@ bool ft8_update_vars_from_config(bool called_at_init_or_restart)
 
 // order matches ft8.autorun_u in FT8.js
 // only add new entries to the end so as not to disturb existing values stored in config
-#define FT4_BAND_IDX 13
 static double ft8_cfs[] = {     // usb carrier/dial freq
     /* FT8 */ 1840, 3573, 5357, 7074,   10136, 14074, 18100, 21074, 24915, 28074, 50313, 40680, 60074,
     /* FT4 */       3575.5,     7047.5, 10140, 14080, 18104, 21140, 24919, 28180, 50318,
+    /* FT8 */ 144174, 222065, 432174, 1296174, 10489540,
+    /* FT4 */ 144150,
 };
 
+// must map to ft8_cfs[] entries
+static u1_t isFT4[] = { 0,0,0,0,0,0,0,0,0,0,0,0,0, 1,1,1,1,1,1,1,1,1, 0,0,0,0,0, 1 };
+
 static const char* ft8_name[] = {
     "160m", "80m", "60m", "40m", "30m", "20m", "17m", "15m", "12m", "10m", "6m", "8m*", "5m*",
-            "80m",        "40m", "30m", "20m", "17m", "15m", "12m", "10m", "6m"
+            "80m",        "40m", "30m", "20m", "17m", "15m", "12m", "10m", "6m",
+    "2m", "1.25m", "70cm", "23cm", "QO-100",
+            "2m",
 };
 
 void ft8_update_spot_count(int rx_chan, u4_t spot_count)
@@ -345,7 +351,7 @@ static void ft8_autorun(int instance, bool initial)
     rx_util_t *r = &rx_util;
     int band = ft8_arun_band[instance]-1;
     double dial_freq_kHz = ft8_cfs[band];
-    bool ft4 = (band >= FT4_BAND_IDX);
+    bool ft4 = (isFT4[band] != 0);
     bool preempt = (ft8_arun_preempt[instance] != ARUN_PREEMPT_NO);
     char *ident_user;
     asprintf(&ident_user, "FT%d-%s", ft4? 4:8, ft8_name[band]);
diff --git a/web/extensions/FT8/FT8.js b/web/extensions/FT8/FT8.js
index 179ea2c..e3dd15c 100644
--- a/web/extensions/FT8/FT8.js
+++ b/web/extensions/FT8/FT8.js
@@ -13,8 +13,8 @@ var ft8 = {
    mode_s: ['FT8', 'FT4'],
 
    freq_s: {
-      'FT8': [ '1840', '3573', '5357', '7074', '10136', '14074', '18100', '21074', '24915', '28074', '50313', '40680', '60074' ],
-      'FT4': [ '3575.5', '7047.5', '10140', '14080', '18104', '21140', '24919', '28180', '50318' ]
+      'FT8': [ '1840', '3573', '5357', '7074', '10136', '14074', '18100', '21074', '24915', '28074', '50313', '40680', '60074', '144174', '222065', '432174', '1296174', '10489540' ],
+      'FT4': [ '3575.5', '7047.5', '10140', '14080', '18104', '21140', '24919', '28180', '50318', '144150' ]
    },
    PASSBAND_LO: 100,
    PASSBAND_HI: 3100,
@@ -35,7 +35,9 @@ var ft8 = {
    autorun_u: [
       'regular use',
       'FT8-160m', 'FT8-80m', 'FT8-60m', 'FT8-40m', 'FT8-30m', 'FT8-20m', 'FT8-17m', 'FT8-15m', 'FT8-12m', 'FT8-10m', 'FT8-6m', 'FT8-8m*', 'FT8-5m*',
-                  'FT4-80m',            'FT4-40m', 'FT4-30m', 'FT4-20m', 'FT4-17m', 'FT4-15m', 'FT4-12m', 'FT4-10m', 'FT4-6m'
+                  'FT4-80m',            'FT4-40m', 'FT4-30m', 'FT4-20m', 'FT4-17m', 'FT4-15m', 'FT4-12m', 'FT4-10m', 'FT4-6m',
+      'FT8-2m', 'FT8-1.25m', 'FT8-70cm', 'FT8-23cm', 'FT8-QO-100',
+                  'FT4-2m'
    ],
 
    PREEMPT_NO: 0,
