diff --git a/extensions/wspr/wspr_main.cpp b/extensions/wspr/wspr_main.cpp
index cbb23b1..39be5c9 100644
--- a/extensions/wspr/wspr_main.cpp
+++ b/extensions/wspr/wspr_main.cpp
@@ -80,7 +80,8 @@ static float window[NFFT];
 #define WSPR_BAND_IWBP 9999
 static double wspr_cfs[] = {
     137.5, 475.7, 1838.1, 3570.1, 3594.1, 5288.7, 5366.2, 7040.1, 10140.2, 14097.1, 18106.1, 21096.1, 24926.1, 28126.1,
-    40680, 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5, 6781.5, 13555.4, WSPR_BAND_IWBP
+    40680, 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5, 6781.5, 13555.4, WSPR_BAND_IWBP,
+    10489569.5
 };
 
 static const char* wspr_name[] = {
diff --git a/web/extensions/wspr/wspr.js b/web/extensions/wspr/wspr.js
index d9f421e..b963f85 100644
--- a/web/extensions/wspr/wspr.js
+++ b/web/extensions/wspr/wspr.js
@@ -37,7 +37,7 @@ var wspr = {
    autorun_u: [
       'regular use', 'LF', 'MF', '160m', '80m_JA', '80m', '60m', '60m_EU',
       '40m', '30m', '20m', '17m', '15m', '12m', '10m',
-      '8m', '6m', '4m', '2m', '440', '1296', 'ISM_6', 'ISM_13', 'IWBP'
+      '8m', '6m', '4m', '2m', '440', '1296', 'ISM_6', 'ISM_13', 'IWBP', 'QO-100'
    ],
 
    PREEMPT_NO: 0,
@@ -970,9 +970,9 @@ function wspr_draw_pie() {
 // dial freq = cf - bfo, cf aka "tx freq"
 // new entries can only be added at end due to limitations with autorun's wspr.autorun_u stored config value
 // 9999 entry for IWBP makes url param &ext=wspr,iwbp work due to freq range check
-var wspr_center_freqs = [ 137.5, 475.7, 1838.1, 3570.1, 3594.1, 5288.7, 5366.2, 7040.1, 10140.2, 14097.1, 18106.1, 21096.1, 24926.1, 28126.1, 40680, 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5, 6781.5, 13555.4, 9999 ];
-var wspr_freqs_s = { 'lf':0, 'mf':1, '160m':2, '80m_ja':3, '80m':4, '60m':5, '60m_eu':6, '40m':7, '30m':8, '20m':9, '17m':10, '15m':11, '12m':12, '10m':13, '8m':14, '6m':15, '4m':16, '2m':17, '440':18, '1296':19, 'ism_6':20, 'ism_13':21, 'iwbp':22 };
-var wspr_freqs_m = [ 'LF', 'MF', '160m', '80m_JA', '80m', '60m', '60m_EU', '40m', '30m', '20m', '17m', '15m', '12m', '10m', '8m', '6m', '4m', '2m', '440', '1296', 'ISM_6', 'ISM_13', 'IWBP' ];
+var wspr_center_freqs = [ 137.5, 475.7, 1838.1, 3570.1, 3594.1, 5288.7, 5366.2, 7040.1, 10140.2, 14097.1, 18106.1, 21096.1, 24926.1, 28126.1, 40680, 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5, 6781.5, 13555.4, 9999, 10489569.5 ];
+var wspr_freqs_s = { 'lf':0, 'mf':1, '160m':2, '80m_ja':3, '80m':4, '60m':5, '60m_eu':6, '40m':7, '30m':8, '20m':9, '17m':10, '15m':11, '12m':12, '10m':13, '8m':14, '6m':15, '4m':16, '2m':17, '440':18, '1296':19, 'ism_6':20, 'ism_13':21, 'iwbp':22, 'qo-100':23 };
+var wspr_freqs_m = [ 'LF', 'MF', '160m', '80m_JA', '80m', '60m', '60m_EU', '40m', '30m', '20m', '17m', '15m', '12m', '10m', '8m', '6m', '4m', '2m', '440', '1296', 'ISM_6', 'ISM_13', 'IWBP', 'QO-100' ];
 
 // freqs on github.com/HB9VQQ/WSPRBeacon are cf - 1.5 kHz BFO (dial frequencies)
 // so we add 1.5 to those to get our cf values (same as regular WSPR wspr_center_freqs above)
@@ -981,8 +981,8 @@ var wspr_cf_IWBP = [ 1838.1, 3570.1, 5288.7, 7040.1, 10140.2, 14097.1, 18106.1,
 //var wspr_cf_IWBP = [ 7040.1, 10140.2, 14097.1 ];
 
 // only one of these is chosen given our 30/32 MHz span
-var wspr_xvtr_center_freqs = [ 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5 ];
-var wspr_xvtr_freqs_s = [ '6m', '4m', '2m', '440', '1296' ];
+var wspr_xvtr_center_freqs = [ 50294.5, 70092.5, 144490.5, 432301.5, 1296501.5, 10489569.5 ];
+var wspr_xvtr_freqs_s = [ '6m', '4m', '2m', '440', '1296', 'QO-100' ];
 
 var wspr_rfreq=0, wspr_tfreq=0;
 var wspr_bfo = 750;
